frame/contracts/src/wasm/runtime.rs

```diff
@@ -705,22 +717,55 @@ where
 		if value_len > max_size {
 			return Err(Error::<E::T>::ValueTooLarge.into())
 		}
-		let mut key: StorageKey = [0; 32];
+
+		let mut key = vec![0; key_len.try_into().map_err(|_| Error::<E::T>::ValueTooLarge)?];
 		self.read_sandbox_memory_into_buf(key_ptr, &mut key)?;
```

Comment:
We are allocating memory based on user input here (`key_len` can be supplied by the contract). Needs to be capped by the maximum size.

```suggestion
                ensure!(key_len <= <E::T>::MaxStorageKeyLen, Error::<E::T>::DecodingFailed);
		let key = self.read_sandbox_memory(key_ptr, key_len)?
```